File: python/Dll.py
from abc import ABCMeta, abstractmethod

class Node:
	
	def __init__(self,before,after,data):
		self.before = before
		self.after = after
		self.data = data

	def getAfter(self):
		return self.after

	def setAfter(self,after):
		self.after = after

	def setBefore(self,before):
		self.before = before

	def getData(self):
		return self.data


class DoublyLinkedListDefinition:
	__metaclass__ = ABCMeta

	@abstractmethod
	def add(self, data):
		pass
	
	@abstractmethod
	def display(self):
		pass

class DoublyLinkedList(DoublyLinkedListDefinition):

	def __init__(self):
		self.head = Node(None,None,None)
		self.tail = Node(None,None,None)
		self.size = 0

	def add(self, data):
		newNode = Node(None, self.head, data)
		'''print "palash>>>>1>>>"+newNode.getData().__str__()'''
		if self.size == 0 :
			self.head = newNode
			'''print "palash>>>>1>>>"+self.head.getData().__str__()'''
		else:
			newNode.setBefore(self.head)
			self.head.setAfter(newNode)
		self.size = self.size + 1

	def display(self):
		i=0
		temp = self.head
		while i < self.size :
			print 'Dll['+i.__str__()+'] # '+temp.getData().__str__()
			temp = temp.getAfter()
			i = i+1

dll = DoublyLinkedList()
dll.add(5)
dll.add(2)
dll.display()

File: java/Dll.java
import java.util.*;

interface DoublyLinkedListInterface<T> {
	boolean	add(T e);//Appends the specified element to the end of this list.
	void display();//print the contents from head to tail
}

class Node <T> {
	T data;
	Node<T> before;
	Node<T> after;

	Node(T data, Node<T> before, Node<T> after){
		this.data=data;
		this.before=before;
		this.after=after;
	}

	void setAfter(Node<T> after){
		this.after=after;
	}

	void setBefore(Node<T> before){
		this.before=before;
	}

	T getData(){
		return this.data;
	}

	Node<T> getAfter(){
		return this.after;
	}
}

class DoublyLinkedListImpl<T> implements DoublyLinkedListInterface<T> {
	Node<T> head;
	Node<T> tail;
	int size=0;

	public boolean add(T data){
		Node<T> newNode = new Node<T>(data,head,null);
		if(head==null) head=newNode;
		else{
			head.setAfter(newNode);
			newNode.setBefore(head);
		}
		size++;
		return true;
	}

	public void display(){
		Node<T> temp=head;
		for(int i=0;i<size;i++){
			if(temp!=null) {
				System.out.println("Node["+i+"] # "+temp.getData());
				temp=temp.getAfter();
			}
		}
	}
}

public class Dll{
	public static void main(String args[]){
		DoublyLinkedListImpl<Object> dll = new DoublyLinkedListImpl<Object> ();
		dll.add("asghasasa");
		dll.add(12);
		dll.display();
	}
}